Adding composition to Shiny app

steffilazerte opened the pull request:
This pull request adds dealing with moose composition to the Shiny app in moosecounter. I'll keep track of my progress, to-do-list and questions in the comments of this pull request. Where I think there's something in the functions of moosecounter that need to be tweaked, I'll open an issue.

## To-do

**Panels to add**
- [x] Explore composition
- [x] Fit composition models
- [x] Estimate composition intervals
- [X] Summaries and downloads

**Documentation**
- [x] Docs...

**Best-practices**
- [x] Tests
- [x] Validate/need checks

steffilazerte committed:
Initial setup for composition

steffilazerte committed:
Be explicit with package versions for VGAM

steffilazerte committed:
Use Total models tab to add Composition models tab

psolymos commented:
Hi @steffilazerte thanks for going over the functions I added. Addressed the 3 issues you have identified.

steffilazerte committed:
Merge branch 'main' of github.com:psolymos/moosecounter into main

steffilazerte committed:
Add composition prediction intervals

steffilazerte committed:
Add shiny progress bars to mc_predict_comp()

steffilazerte committed:
Cleanup

steffilazerte committed:
Standardize parameter names

steffilazerte committed:
Catch models which did not converge

steffilazerte committed:
Catch cells with no variability

steffilazerte committed:
Fix names in comp and allow multiple total models (but not comp models)

steffilazerte committed:
Warn on comp PI when total or comp models lists change

steffilazerte committed:
housekeeping

steffilazerte commented:
Hi Peter, 
I've got the first pass of the main panels finished. A couple of things to note:
- I renamed parameters to line up with total_ and comp_ subsections
- I added a catch to total models when the model didn't converge, which will pop up in the error messages. 
- I added a catch to the map when the column to plot had no varaiability

Questions
- Should the composition exploration stay under composition, or move to the exploration section (ie. with Univariate and Multivariate)
- When it comes to documentation, what did you have in mind? Docs relating to the function documentation in moosecounter, or more related to a tutorial on how to use the Shiny App?

psolymos commented:
Thanks @steffilazerte I will check your changes in this PR.

> Should the composition exploration stay under composition, or move to the exploration section (ie. with Univariate and Multivariate)

Good question. Now that I think about it, it can go under the Exploration section as a 3rd subsection "Composition".

I think "Univeriate" and "Multivariate" can stay as is in the side bar (but these refer to the total moose/cow models), but the heading can be more specific: "Multivariate Exploration, Total Abundance" etc.

> When it comes to documentation, what did you have in mind? Docs relating to the function documentation in moosecounter, or more related to a tutorial on how to use the Shiny App?

It is 2-fold:

- the docs can be improved a bit, I can take a stab at it
- as part of the app I am thinking of a short tutorial that explain the tabs and what the user is supposed to do, also links to downloadable example files so that they can have a data set to work with (I used markdown for this in the past: e.g. `includeMarkdown("docs.md")`)

psolymos commented on R/comp.R:
```diff
@@ -209,6 +209,12 @@ mc_predict_comp <- function(total_model_id, comp_model_id,
   b <- 1
+
+  # Progress bars setup
+  if(requireNamespace("shiny") && shiny::isRunning()) {
```
Of course! This brings back memories :)

psolymos commented on inst/shiny/server.R:
```diff
@@ -565,4 +577,247 @@ server <- function(input, output, session) {
+    req(input$comp_model_id)
+    comp_models_list$m[[input$comp_model_id]] <- list(var = input$comp_model_var)
+  }) %>%
+    bindEvent(input$comp_model_add)
```
Nice use of `bindEvent`

steffilazerte replied:
Actually thanks for the reminder, I was planning on updating the eventReactives and observeEvents to use bindEvent, as I think that's the standard now.

psolymos commented:
The select/selectize options for total and comp are different:
<img width="450" alt="Screen Shot 2022-02-16 at 8 11 08 PM" src="https://user-images.githubusercontent.com/831636/154397951-e83693b9-36e5-4df6-989f-8c2d3580011b.png">

psolymos commented:
Hey @steffilazerte , thank you for your work on the app, again! I added a screenshot above about a little inconsistency that I noted. Other than that, the app works as expected. Nice work!

I was thinking about your previous question of where to put the univariate/multivariate exploration subsections. Given that the comp exploration is now part of the Composition section, would it make sense to move univariate/multivariate under Total? I think it would make sense.

Regarding the Summary, I was thinking and I am not sure if cell level estimates are that useful. A download option would be useful (right now we have the output from `pred_density_moose_CPI()` that is not much). So I will reach out to the biologists and ask them what they would like to have, if anything. No need to add anything more at the moment, and the placeholders are fine.

steffilazerte commented:
Hi @psolymos, unfortunately I'm not sure we can make the two menus consistent, as I think the differences reflect the fact that one can select multiple models (total) and the other only one (comp). I'll see what I can do!

I'll look into the documentation next, then. Let me know when you've got more ideas for the summary :grin:

psolymos commented:
> the differences reflect the fact that one can select multiple models (total) and the other only one (comp)

OK, that make total sense.

psolymos commented:
Hi @steffilazerte , here is what I think about the composition PI exploration:

I made some changes exposing a function `subset_CPI_data` that I overlooked, which is useful in the console (not so much in the app). I also added few lines to the workflow:

```R
# print composition results based on all rows
pred_density_moose_CPI(CPI)

## print composition results based on a subset of rows
pred_density_moose_CPI(subset_CPI_data(CPI, ss = 1:100))

## summaries of the cohort PI's
head(CPI$cells)
```

We can add the `CPI$cells` info to the _Composition summary_ tab, along with the `"SU_ID"` column. 

I am not quite sure if the `subset_CPI_data` is useful here, subsetting can be done on a downloadable file. Let's assume this situation for now and add another file download similar to the `get_xlslist` reactive and the `boot_download` download handler. If you can take a 1st stab at it I can then edit and make sure we have all the info we need.

Thanks!

steffilazerte committed:
Use bindEvent() rather than observeEvent() or eventReactive()

steffilazerte committed:
Merge branch 'main' of github.com:psolymos/moosecounter into main

steffilazerte committed:
Move uni/multivariate explorations under total

steffilazerte committed:
Clarify created observe events

steffilazerte committed:
Standardize names/titles

steffilazerte committed:
Rearrange and add boot strap table to Comp PIs

steffilazerte committed:
Add summary and download option

steffilazerte committed:
Add placeholder height

steffilazerte commented:
@psolymos I think I've added what you wanted for the summaries, (`CPI$cells`) and the download option. I've also rearranged the Composition > PI tab and added the bootstrap table like for the Total > PI tab. 

Did you want the option to subset the data for Comp > Summary? I wasn't clear which situation you were assuming :grin: Next for me are docs!

psolymos commented:
Hey @steffilazerte I can't see the most recent commits, and the bootstrap table. Did you push all what you have?

steffilazerte commented:
Argh, nope! pushed now

steffilazerte committed:
Merge branch 'main' of github.com:psolymos/moosecounter into main

psolymos commented:
- Let's wait for feedback on Comp > Summary, but I expect this to be minor/straightforward
- Docs for the app [SL] and package [PS] should be priority

When I look at the Comp Summary, the box does not encapsulate the table for some reason.

steffilazerte commented:
> Docs for the app [SL] and package [PS] should be priority

Sounds good!

> When I look at the Comp Summary, the box does not encapsulate the table for some reason.

I've had this problem on other apps, I'll look into it, though

steffilazerte committed:
Make sure box auto-adjusts to height of contents, but use min-height div to ensure enough room for messages.

steffilazerte commented:
Ah, got the second point. I set the height of the box to account for the "Load data first" message, but that prevented it from auto adjusting. So instead I used a div with min-height. Hope this helps on my other app too!

steffilazerte committed:
Get rid of duplicate data message

steffilazerte committed:
Documentation start

steffilazerte commented:
I started on the documentation this morning. It's still rough and incomplete, but I wanted to check in with you that this is somewhat what you had in mind. You may need to make some corrections if I've used incorrect terminology or misunderstood the purpose of a section, but in general, is this okay? I could include screenshots as well, but I'm not sure how useful that is (they need to be updated if the app is updated, and if the docs are in the app itself, is it redundant?)

Let me know if you had envisioned more/less details or a different setup

psolymos commented:
Really good start. Just what I had in mind w.r.t. the level of detail. I agree screenshots might be a bit early. Thanks!

steffilazerte committed:
Complete docs

steffilazerte commented:
I think I'm done with the documentation. I ended up making it an Rmd file that you need to compile into the md file (so I could get the list of response and metadata variables). 

One thing I haven't added that might be useful is a list of output variables for the Total and Composition PIs and what they mean. I think I might need a bit of help with that, since I don't have a great understanding of all the outputs 😅 . But if you think that would be valuable, you could add a csv (or similar) table which I could then pull on. 

Let me know if you think any of the sections need more elaboration or if there's something else I should include!

steffilazerte committed:
Update and add tests

steffilazerte commented:
So I think I'm done, except for any tweaks you ask for. Let me know what you need!

psolymos commented:
I like the summary. Right amount of detail.
